src/wutil.cpp

```diff
@@ -711,6 +711,78 @@ double fish_wcstod(const wchar_t *str, wchar_t **endptr) {
     return std::wcstod(str, endptr);
 }
 
+/// Like wcstod(), but allows underscore dividers between digits in the input string.
+double fish_wcstod_underscores(const wchar_t *str, wchar_t **endptr) {
+    // OpenBSD's strtod/wcstod do not support hex.
+    static const bool allow_hex = (fish_wcstod(L"0x1", nullptr) == 1);
+    const wchar_t *orig = str;
+    while (iswspace(*str)) str++;
+    std::wstring pruned;
+    auto take_char = [&pruned, &str]() {
+        if (*str != L'_') pruned.push_back(*str);
+        str++;
+    };
+    auto is_sign = [](wchar_t c) { return c == L'+' || c == L'-'; };
+    if (is_sign(*str)) take_char();
+    if (!wcscasecmp(str, L"infinity") || !wcscasecmp(str, L"nan")) {
+        return fish_wcstod(orig, endptr);
+    }
+    bool hex = false;
+    if (allow_hex && !wcsncmp(str, L"0x", 2)) {
+        take_char();
+        take_char();
+        hex = true;
+    }
+    auto is_mantissa_digit = [hex](wchar_t c) -> int { return hex ? iswxdigit(c) : iswdigit(c); };
+    bool got_radix = false;
+    bool got_digit = false;
+    while (is_mantissa_digit(*str) || *str == L'.' || *str == L'_') {
+        if (*str == L'_') {
+            if (str == orig || !is_mantissa_digit(*(str - 1)) || !is_mantissa_digit(*(str + 1))) {
+                break;
+            }
+        } else if (*str == L'.') {
+            if (got_radix) break;
+            got_radix = true;
+        } else {
+            got_digit = true;
+        }
+        take_char();
+    }
+    if (!got_digit) {
+        if (hex) {
+            str = orig;
+            while (*str != L'x') str++;
+            // hex = false; // This cannot have an effect in the rest of the function, so skip it.
+        } else {
+            if (endptr) *endptr = const_cast<wchar_t *>(orig);
+            return 0.0;
+        }
+    }
+    auto is_exp_char = [hex](wchar_t c) {
+        if (hex) {
+            return c == L'P' || c == L'p';
```

Comment:
I'm not sure this isn't over-specified. Wouldn't it be simpler to read until a separator (`,`, `)`, whitespace, end of the string)?

Overreading shouldn't be an issue because wcstod can just return the end pointer, which we can then offset by the number of `_`.

Or is there a reason you're doing it like this?

Reply:
I.e.

1. Create the pruned string, counting the number of `_`
2. Call wcstod, passing a second endptr
3. Set the real endptr to start of real string + offset of the second endptr + number of `_`
4. Return wcstod value

Reply:
This has the advantage that, if wcstod behaves differently than we think, we don't get weird double-counting or skipping, but have the end at the place it gives us.

Reply:
It's not quite as simple as consuming all non-underscore characters until a separator and counting how many underscores you dropped in the process for the eventual offset.

First, if you don't implement real parsing logic, then you will convert `1_1.2_2.3_3` to `11.22.33` and `wcstod` will only parse until the second decimal. Which is fine, but you need to know that you should only advance your own `endptr` with an offset assuming you dropped two underscores, not all three.

Second, the situation depends on what rules we want to use for where underscores can appear. If we want to have nontrivial restrictions, then I think we basically need to implement the full parsing logic. For example, deciding whether an `a` is an end-of-parse or a hex digit requires knowing whether the string began with `0x`.

In my mind, originally I was planning to have some rules about where underscores could appear. But if we are ok with being very permissive (as in #8611) that simplifies things and we can get away without a full parse.

I figured that if I was going to do a full parse anyway, then I would just implement the parse corresponding to the C99 spec -- it turns out it's not really that bad. The `allow_hex` runtime feature flag also is not that bad IMO.

It's true that we are dead-reckoning (or praying) in this case that our parse is going to be equal to the system `wcstod` parse, which I agree feels pretty uncomfortable. But if that is not the case then either there is a bug in our parse, or the system `wcstod` is non-conformant. I didn't implement the `endptr` offsetting using the offset we get from the system `wcstod` because (1) if there are no bugs and the system function is conformant then I should be able to assert that it's equal to the end of the pruned string anyway, (2) the fallback logic for offsetting based on number of underscores dropped is not completely trivial (because of the `1_1.2_2.3_3` case above, it's not enough to count), (3) if we have a bug in the parser or the system stdlib function is non-conformant then I didn't see a graceful recovery being likely.

Anyway I do think the new version is better and probably more robust.